libs/libc/unistd/lib_getoptvars.c

```diff
@@ -0,0 +1,153 @@
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <unistd.h>
+#include <pthread.h>
+
+#include "unistd.h"
+#include "libc.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define INVALID_KEY    CONFIG_TLS_NELEM
+#define VALID_KEY(key) ((key) >= 0 && key < CONFIG_TLS_NELEM)
+
+/* Configuration.  Currently no thread-specific data destructors. */
+
+#undef  HAVE_DESTRUCTOR
+#define HAVE_CLEANUP   1
+#define HAVE_TLSVARS   1
+
+/* Check if we have some way to avoid a memory leak */
+
+#if defined(CONFIG_DISABLE_PTHREAD) || !defined(CONFIG_PTHREAD_CLEANUP)
+#  undef HAVE_CLEANUP
+#endif
+
+/* Don't use TLS if we cannot avoid the memory leak or if TLS is disabled */
+
+#if (!defined(HAVE_DESTRUCTOR) && !defined(HAVE_CLEANUP)) || \
+    CONFIG_TLS_NELEM == 0
+#  undef HAVE_TLSVARS
+#endif
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+#ifdef HAVE_TLSVARS
+static pthread_key_t g_getopt_key = (pthread_key_t)INVALID_KEY;
+
+#else
+/* Fallback to avoid memory leak when there is no destructor and no cleanup
+ * or when there is no TLS.
+ */
+
+FAR struct getopt_s g_getopt_vars =
+{
+  NULL,
+  0,
+  1,
+  '?'
+};
+#endif
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+#ifdef HAVE_TLSVARS
+void _getopt_destructor(FAR void *getoptvars)
+{
+  lib_free(getoptvars);
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: _getoptvars
+ *
+ * Description:
+ *   Returns a pointer to to the thread-specific getopt() data.
+ *
+ ****************************************************************************/
+
+FAR struct getopt_s *_getoptvars(void)
+{
+#ifdef HAVE_TLSVARS
+  FAR struct getopt_s *go;
+  int ret;
+
+  /* Check if a key has been assigned */
+
+  if (!VALID_KEY(g_getopt_key))
+    {
+      ret = pthread_key_create(&g_getopt_key, _getopt_destructor);
+      if (ret != OK)
+        {
+          return NULL;
+        }
+
+      /* Allocate getopt() variables */
+
+      go = (FAR struct getopt_s *)lib_malloc(sizeof(struct getopt_s));
+      if (go == NULL)
+        {
+          pthread_key_delete(g_getopt_key);
+          return NULL;
+        }
```

Comment:
same as above. there's no good backup if this happens

Reply:
Done